### PR TITLE
Issue #46: Allow required and optional parameters to be specified whi…

### PR DESCRIPTION
**Pull request by jeffmaury**
…le creating a server

- Add optional parameter attributes to server creation methods

Signed-off-by: Jeff MAURY <jmaury@redhat.com>

### CODE REVIEW

File: src/protocol/protocol.ts [hunk 1]
```diff
@@ -33,6 +33,16 @@ export namespace Protocol {
         params: ServerAttributes;
     }
 
+    export interface Attributes {
```
Comment (jrichter1) [inline src/protocol/protocol.ts]: both these interfaces are already declared in this file, no need to add them anew

File: src/util/serverModel.ts [hunk 2]
```diff
@@ -41,21 +41,22 @@ export class ServerModel {
      * event to see when the server creation is finished
      * @param path path to the server's root directory
      * @param id unique identifier for the newly created server
+     * @param attributes optional extra server attributes
      * @param timeout timeout in milliseconds
      */
-    async createServerFromPathAsync(path: string, id: string, timeout: number = 2000): Promise<Protocol.Status> {
+    async createServerFromPathAsync(path: string, id: string, attributes?: any, timeout: number = 2000): Promise<Protocol.Status> {
```
Comment (jrichter1) [inline src/util/serverModel.ts]: Please don't use the any type. At least use object if there is no more precise matching available.

**Comment by jeffmaury** [comment]
@jrichter1 dare to review ?

**Comment by jeffmaury** [comment]
@jrichter1 dare to review (bis)

**Comment by jrichter1** [comment]
Apologies, the notification must have slipped under my radar, I'll get to it right away.

File: src/client.ts [hunk 3]
```diff
@@ -162,14 +162,14 @@ export class RSPClient {
      * @param id unique identifier for the newly created server
      * @param timeout timeout in milliseconds
      */
-    createServerSync(pathOrBean: string | Protocol.ServerBean, id?: string, timeout: number = 2000): Promise<Protocol.ServerHandle> {
+    createServerSync(pathOrBean: string | Protocol.ServerBean, id?: string, attributes?: any, timeout: number = 2000): Promise<Protocol.ServerHandle> {
```
Comment (jrichter1) [inline src/client.ts]: Please add a type definition for attributes like in the createServerFrom* methods.

File: src/client.ts [hunk 4]
```diff
@@ -181,14 +181,14 @@ export class RSPClient {
      * @param id unique identifier for the newly created server
      * @param timeout timeout in milliseconds
      */
-    createServerAsync(pathOrBean: string | Protocol.ServerBean, id?: string, timeout: number = 2000): Promise<Protocol.Status> {
+    createServerAsync(pathOrBean: string | Protocol.ServerBean, id?: string, attributes?: any, timeout: number = 2000): Promise<Protocol.Status> {
```
Comment (jrichter1) [inline src/client.ts]: Please add a type definition for attributes as well.

File: test/client-test.ts [hunk 5]
```diff
@@ -251,7 +251,7 @@ describe('RSP Client', () => {
 
             const result = await client.createServerSync('path', 'id');
 
-            expect(modelStub.createServerFromPath).calledWith('path', 'id', defaultTimeout);
+            expect(modelStub.createServerFromPath).calledWith('path', 'id', undefined, defaultTimeout);
```
Comment (jrichter1) [inline test/client-test.ts]: These methods should be tested with some attributes as well, so we can see if they are passed correctly. Either rewrite these test cases to use attributes, or create separate. I don't mind either :)

**Pull request merged**